File: tests/rpc/withdraw-all.ts
import { SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { Ctx, Member } from "../ctx/ctx";

export async function withdrawalAllRPC(ctx: Ctx, member: Member) {
    await ctx.program.methods.withdrawAll()
    .accounts({
        factory: ctx.PDAS.factory.key,
        beneficiary: member.beneficiary.publicKey,
        beneficiaryTokenAccount: member.beneficiaryStakeVault,
        member: member.key,
        vaultFree: member.vaultFree,
        systemProgram: SystemProgram.programId,
        tokenProgram: TOKEN_PROGRAM_ID,
    })
    .signers([member.beneficiary])
    .rpc();
}

File: programs/staking/src/context/withdraw_all.rs
use anchor_lang::prelude::*;
use crate::account::*;
use anchor_spl::token::{self, TokenAccount, Token};

#[derive(Accounts)]
pub struct WithdrawAll<'info> {
    #[account(
        seeds = [Factory::PDA_SEED],
        bump = factory.bump,
    )]
    pub factory: Account<'info, Factory>,
    #[account(
        seeds = [
            beneficiary.to_account_info().key.as_ref(),
            factory.to_account_info().key.as_ref(),
        ],
        bump = member.bump,
        has_one = vault_free,
    )]
    pub member: Account<'info, Member>,
    #[account(mut)]
    pub vault_free: Box<Account<'info, TokenAccount>>,
    #[account(mut)]
    pub beneficiary: Signer<'info>,
    #[account(mut)]
    pub beneficiary_token_account: Box<Account<'info, TokenAccount>>,
    pub system_program: Program<'info, System>,
    pub token_program: Program<'info, Token>,
}

impl<'info> WithdrawAll<'info> {
    pub fn transfer_free_vault_tokens_to_beneficiary(&self) -> Result<()> {
        let amount = self.vault_free.amount;

        let seeds = &[
            self.beneficiary.to_account_info().key.as_ref(),
            self.factory.to_account_info().key.as_ref(),
            &[self.member.bump]
        ];

        token::transfer(
            CpiContext::new_with_signer(
                self.token_program.to_account_info(),
                token::Transfer { 
                    from: self.vault_free.to_account_info(), 
                    to: self.beneficiary_token_account.to_account_info(), 
                    authority: self.member.to_account_info() 
                },
                &[&seeds[..]],
            ),
            amount
        )
    }
}
